### PR TITLE
5330: Filtering incidents by tags

### PR DESCRIPTION
**Pull request by karthiksbhat**
Server-side code for filtering incidents by tags.
If a single tag is provided, all incidents with that tag are shown.
If multiple tags are provided, all incidents that contain `at least 1` of those tags are shown.

### CODE REVIEW

File: models/incident.js [hunk 1]
```diff
@@ -123,25 +123,35 @@ Incident.queryIncidents = function(query, page, options, callback) {
...
   else delete filter.locationName;
 
+  // Checking for multiple tags in incident
+  if (query.tags) {
+    var tags = [];
+    for (var i = 0; i < query.tags.length; i++) {
+      tags[i] = query.tags[i];
+    }
+    filter.tags = {};
+    filter.tags.$in = tags;
```
Comment (pgarrison) [inline models/incident.js]: I would suggest `filter.tags = { $in: query.tags };`, or if it's important to get a copy, `filter.tags = { $in: query.tags.slice() };`.

File: test/backend/models.query.incident-query.test.js [hunk 2]
```diff
@@ -15,10 +15,12 @@ describe('Incident query attributes', function() {
...
-        { title: 'The quick brown fox', veracity: null, closed: true },
-        { title: 'The slow white fox', veracity: true, closed: false },
-        { title: 'The quick brown chicken', veracity: null, closed: false },
-        { title: 'The brown quick fox', veracity: false, closed: true }
+        { title: 'The quick brown fox', veracity: null, closed: true, tags: ['hello', 'world'] },
+        { title: 'The slow white fox', veracity: true, closed: false, tags: ['hello'] },
+        { title: 'The quick brown chicken', veracity: null, closed: false, tags: ['world'] },
+        { title: 'The brown quick fox', veracity: false, closed: true, tags: ['helloworld', 'wellohorld', 'foobar'] },
+        { title: 'The fox that was slow and brown', veracity: false, closed: true, tags: ['foobar', 'hello'] },
+        { title: 'The slow that was fox and brown', veracity: false, closed: true, tags: ['wellohorld', 'hello'] }
```
Comment (pgarrison) [inline test/backend/models.query.incident-query.test.js]:
```javascript
{
  title: 'The slow that was fox and brown',
  veracity: false,
  closed: true,
  tags: ['wellohorld', 'hello']
}
```
?

File: test/backend/models.query.incident-query.test.js [hunk 3]
```diff
@@ -93,6 +95,28 @@ describe('Incident query attributes', function() {
...
+    (new IncidentQuery({ veracity: 'Confirmed false', status: 'closed', tags: ['wellohorld', 'foobar'] })).run(function(err, incidents) {
+      if (err) return done(err);
+      expect(incidents).to.have.keys(['total', 'results']);
+      expect(incidents.total).to.equal(3);
+      expect(incidents.results).to.be.an.instanceof(Array);
+      expect(incidents.results).to.have.length(3);
+      done();
+    });
+  });
+
```
Comment (pgarrison) [inline test/backend/models.query.incident-query.test.js]: Consider
```javascript
var foo = function (tags, n) {
  return function(done) {
    new IncidentQuery({ 
      veracity: 'Confirmed false', 
      status: 'closed', 
      tags: tags
    }).run(function(err, incidents) {
      if (err) return done(err);
      expect(incidents).to.have.keys(['total', 'results']);
      expect(incidents.total).to.equal(n);
      expect(incidents.results).to.be.an.instanceof(Array);
      expect(incidents.results).to.have.length(n);
      done();
    });
  };
};
```

File: test/backend/models.query.incident-query.test.js (hunk 2 above)
Comment (karthiksbhat) [inline test/backend/models.query.incident-query.test.js]: I just followed the format of what was already there. I'll make this change, as it looks better.

**Pull request merged**